File: core/src/main/java/dev/bee/kanjianki/core/FlashcardGesturePolicy.java
package dev.bee.kanjianki.core;

public final class FlashcardGesturePolicy {
    private static final float HORIZONTAL_DOMINANCE = 1.25f;

    private FlashcardGesturePolicy() {
    }

    public static Decision release(
            float startX,
            float startY,
            float endX,
            float endY,
            int touchSlop,
            int minimumSwipeDistance,
            boolean answerRevealed
    ) {
        float dx = endX - startX;
        float dy = endY - startY;
        float absX = Math.abs(dx);
        float absY = Math.abs(dy);
        int safeTouchSlop = Math.max(0, touchSlop);
        if (absX <= safeTouchSlop && absY <= safeTouchSlop) {
            return answerRevealed ? Decision.none() : Decision.reveal();
        }
        int swipeThreshold = Math.max(Math.max(0, minimumSwipeDistance), safeTouchSlop * 6);
        if (absX >= swipeThreshold && absX > absY * HORIZONTAL_DOMINANCE && answerRevealed) {
            return Decision.review(dx > 0 ? StudyRatings.GOOD : StudyRatings.AGAIN);
        }
        return Decision.none();
    }

    public static final class Decision {
        public enum Action {
            NONE,
            REVEAL,
            REVIEW
        }

        public final Action action;
        public final String rating;

        private Decision(Action action, String rating) {
            this.action = action;
            this.rating = rating;
        }

        public static Decision none() {
            return new Decision(Action.NONE, "");
        }

        public static Decision reveal() {
            return new Decision(Action.REVEAL, "");
        }

        public static Decision review(String rating) {
            return new Decision(Action.REVIEW, rating);
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/FlashcardGesturePolicy.kt
package dev.bee.kanjianki.core

import kotlin.math.abs
import kotlin.math.max

object FlashcardGesturePolicy {
    private const val HORIZONTAL_DOMINANCE = 1.25f

    @JvmStatic
    fun release(
        startX: Float,
        startY: Float,
        endX: Float,
        endY: Float,
        touchSlop: Int,
        minimumSwipeDistance: Int,
        answerRevealed: Boolean,
        swipeEnabled: Boolean = true,
    ): Decision {
        val dx = endX - startX
        val dy = endY - startY
        val absX = abs(dx)
        val absY = abs(dy)
        val safeTouchSlop = max(0, touchSlop)
        if (absX <= safeTouchSlop && absY <= safeTouchSlop) {
            return if (answerRevealed) Decision.none() else Decision.reveal()
        }
        val swipeThreshold = max(
            max(0, minimumSwipeDistance),
            saturatingMultiplyNonNegative(safeTouchSlop, 6),
        )
        if (swipeEnabled && absX >= swipeThreshold && absX > absY * HORIZONTAL_DOMINANCE && answerRevealed) {
            return Decision.review(if (dx > 0) StudyRatings.GOOD else StudyRatings.AGAIN)
        }
        return Decision.none()
    }

    class Decision private constructor(
        @JvmField val action: Action,
        @JvmField val rating: String,
    ) {
        enum class Action {
            NONE,
            REVEAL,
            REVIEW,
        }

        companion object {
            @JvmStatic
            fun none(): Decision = Decision(Action.NONE, "")

            @JvmStatic
            fun reveal(): Decision = Decision(Action.REVEAL, "")

            @JvmStatic
            fun review(rating: String): Decision = Decision(Action.REVIEW, rating)
        }
    }
}
